src/dft/src/architect/ScanArchitectHeuristic.cpp

```diff
@@ -32,21 +32,97 @@
 
 #include "ScanArchitectHeuristic.hh"
 
+#include <boost/geometry.hpp>
+#include <boost/geometry/geometries/register/point.hpp>
+#include <boost/geometry/index/rtree.hpp>
 #include <iostream>
 #include <sstream>
 
 #include "ClockDomain.hh"
 #include "ScanArchitect.hh"
 
+namespace bg = boost::geometry;
+namespace bgi = boost::geometry::index;
+
 namespace dft {
 
 ScanArchitectHeuristic::ScanArchitectHeuristic(
     const ScanArchitectConfig& config,
-    std::unique_ptr<ScanCellsBucket> scan_cells_bucket)
-    : ScanArchitect(config, std::move(scan_cells_bucket))
+    std::unique_ptr<ScanCellsBucket> scan_cells_bucket,
+    utl::Logger* logger)
+    : ScanArchitect(config, std::move(scan_cells_bucket)), logger_(logger)
 {
 }
 
+void ScanArchitectHeuristic::OptimizeScanWirelength(
+    std::vector<std::unique_ptr<ScanCell>>& cells)
+{
+  // Nothing to order
+  if (cells.empty()) {
+    return;
+  }
+  // No point running this if the cells aren't placed yet
+  for (const auto& cell : cells) {
+    if (!cell->isPlaced()) {
+      return;
+    }
+  }
+  // Define the starting node as the upperleftmost
+  size_t start_index = 0;
+  int64_t lowest_dist = std::numeric_limits<int64_t>::max();
+  for (size_t i = 0; i < cells.size(); i++) {
+    int64_t dist = cells.at(i)->getOrigin().x() + cells.at(i)->getOrigin().y();
+    if (dist < lowest_dist) {
+      start_index = i;
+      lowest_dist = dist;
+    }
+  }
+  // Update the index
+  typedef bg::model::point<int, 2, bg::cs::cartesian> bg_point;
+  bgi::rtree<std::pair<bg_point, size_t>, bgi::rstar<4>> rtree;
+  std::pair<bg_point, size_t> cursor;
+  for (size_t i = 0; i < cells.size(); i++) {
+    auto origin = cells.at(i)->getOrigin();
+    auto pair = std::make_pair(bg_point(origin.x(), origin.y()), i);
+    if (i == start_index) {
+      cursor = pair;
+    }
+    rtree.insert(pair);
+  }
+  // Search nearest neighbours
+  std::vector<std::unique_ptr<ScanCell>> result;
+  // Max number of cells we assume could be placed at the same point (hopefully,
+  // after DP, this is never more than one)
+  const int kMaxCellsToSearch = 10;
+
+  while (true) {
+    result.emplace_back(std::move(cells.at(cursor.second)));
+    if (result.size() >= cells.size()) {
+      break;
+    }
+    // Search for next nearest
+    auto it = rtree.qbegin(bgi::nearest(cursor.first, kMaxCellsToSearch));
+    auto next = cursor;
+    while (it != rtree.qend()) {
+      next = *it++;
```

Comment:
This is creating an implicit temporal copy of the iterator. Please try to avoid postfix ++ as much as possible.